### DISCUSSION TITLE
Provide mechanism to load JSON as a dictionary

### DISCUSSION THREAD

**Discussion by virtuald** [mesonbuild/meson#13649] [Ideas]
I'd like an easy way to create a Meson dictionary from a JSON-formatted string.

The use case I'm thinking of is to make it easier to use external scripts to generate structured data that can be used in a build script. For example doing something like this is currently very hard to do if you need more than just key/value data:

```
prog = find_program('prog')
kv = run_command(prog, capture: true).stdout().fromJsonString()

foreach k, v in kv
    foreach ik, iv in v
        ...
    endforeach
endforeach

```

I'm not sold on the method name, it could also be `asDict` or `fromJson` or various other options.

I'd be happy to make a PR for this if there's support for it.

---

An alternative approach would be to add a `json` module like the `keyval` module, but that seems a bit more clunky.

**Comment by eli-schwartz** [comment]
The keyval module was designed by and for projects using configurations similar to the linux kernel Kconfig (the module was originally called the "kconfig" module). It's assumed to be part of your source code and users edit it externally, then run meson. If the keyval file being loaded does change for any reason, it triggers a reconfigure.

run_command is basically "unclean" as it were. Using it goes outside of meson's predictions of what your build looks like, we cannot reconfigure if the expected output changes, etc. There are still lots of scenarios where it's vital for use and there is no replacement, but I'm nervous about anything that looks like trying to define the build graph in this manner, and I feel uncomfortable making changes that result in it looking "too attractive" to the unsuspecting. Still, maybe another maintainer has a different opinion. /cc @dcbaker @jpakkane @xclaesse.

**Reply by virtuald** [reply]
That's fair.

What about adding a `readJSON` method to `fs` instead? That would allow meson to reconfigure if the read file changed? Then to satisfy the original scenario I was thinking about, you could have a generator or some other thing write the json to file.

**Reply by eli-schwartz** [reply]
I am not sure I fully understand the suggestion, but it sounds interesting and may be food for thought.

**Reply by dcbaker** [reply]
I share discomfort in the idea of directly interpreting the output of a command as JSON data, I have thought vaguely about adding `fs.read_json()`, but I've never had a concrete use for it, so it's never gotten far in my mind.

**Comment by virtuald** [comment]
Hm. What if meson had an `unsafe` module, where things like this could go?

Then it would be clear that the user really should avoid using it. It does feel like reading through the various issues/SO posts, that there's lots of features that people ask for that are.. sometimes reasonable in extreme circumstances, but most people shouldn't do them.

Honestly, you could add things like glob to an unsafe module too... haha.

**Comment by dvdhrm** [comment]
I would prefer using structured data over `strip()` and `split()` which meson currently provides on strings. So whenever I resort to `split()`, I much much rather prefer using `from_json()` and ensure that the source data is encoded as JSON. Any list, array, environment, enumeration, etc., it would be so much easier if we avoided custom string manipulation and just instead have the option to treat it as structured JSON data.

So something like: `string.parse_json() -> any` which parses a string as JSON and returns one of string, int, array, or dict.

Ideally it comes with its counterpart: `any.as_json() -> string` (implement on the mentioned types)

I would love to drop all the strip() calls, split() calls, and other string manipulations, and instead convert everything over to these.
